### PR TITLE
ENH: Issue warning when NIFTI IO coerces matrix to orthogonal.

### PR DESCRIPTION
**Pull request by zivy**
addresses #3476

### CODE REVIEW

File: Modules/IO/NIFTI/src/itkNiftiImageIO.cxx [hunk 1]
```diff
@@ -2151,6 +2151,24 @@ NiftiImageIO::SetNIfTIOrientationFromImageIO(unsigned short origdims, unsigned s
...
+  vnl_matrix<float>  imagemat(matdim, matdim);
+  for (unsigned c = 0; c < matdim; ++c)
+  {
+    auto col = this->GetDirection(c);
+    for (unsigned r = 0; r < matdim; ++r)
+    {
+      imagemat[r][c] = col[r];
+    }
+  }
+  auto candidate_identity = imagemat * imagemat.transpose();
```
Comment (jhlegarreta) [inline Modules/IO/NIFTI/src/itkNiftiImageIO.cxx:2166]: Underscores are strongly discouraged in C++ as per the ITK style guidelines: https://itk.org/ItkSoftwareGuide.pdf (section C.6)

File: Modules/IO/NIFTI/src/itkNiftiImageIO.cxx [hunk 2]
```diff
@@ -2151,6 +2151,24 @@ NiftiImageIO::SetNIfTIOrientationFromImageIO(unsigned short origdims, unsigned s
   mat44 matrix =
     nifti_make_orthog_mat44(dirx[0], dirx[1], dirx[2], diry[0], diry[1], diry[2], dirz[0], dirz[1], dirz[2]);
   matrix = mat44_transpose(matrix);
+  // Check if matrix is orthogonal and issue a warning if it was
+  // coerced to orthogonal. Use same epsilon as used in SetImageIOOrientationFromNIfTI
+  const unsigned int matdim(this->GetDirection(0).size());
```
Comment (dzenanz) [inline Modules/IO/NIFTI/src/itkNiftiImageIO.cxx:2156]: I will further suggest renames: `matdim` -> `matDim` or `matrixDim`.

File: Modules/IO/NIFTI/src/itkNiftiImageIO.cxx [hunk 3]
```diff
@@ -2151,6 +2151,24 @@ NiftiImageIO::SetNIfTIOrientationFromImageIO(unsigned short origdims, unsigned s
...
+  vnl_matrix<float>  imagemat(matdim, matdim);
```
Comment (dzenanz) [inline Modules/IO/NIFTI/src/itkNiftiImageIO.cxx:2157]: `imagemat` -> `imageMat` or `dirMat` or `imageDir` or `direction`

File: Modules/IO/NIFTI/src/itkNiftiImageIO.cxx (hunk 1 above)
Comment (zivy) [reply Modules/IO/NIFTI/src/itkNiftiImageIO.cxx:2166]: Prefer to keep this as it is consistent with the naming in `SetImageIOOrientationFromNIfTI` .

File: Modules/IO/NIFTI/src/itkNiftiImageIO.cxx [hunk 4]
```diff
@@ -2151,6 +2151,24 @@ NiftiImageIO::SetNIfTIOrientationFromImageIO(unsigned short origdims, unsigned s
...
+  for (unsigned c = 0; c < matdim; ++c)
+  {
+    auto col = this->GetDirection(c);
+    for (unsigned r = 0; r < matdim; ++r)
+    {
+      imagemat[r][c] = col[r];
+    }
+  }
+  auto candidate_identity = imagemat * imagemat.transpose();
+  if (!candidate_identity.is_identity(1.0e-4))
```
[the remark this answers is not in the document]
Comment (zivy) [reply Modules/IO/NIFTI/src/itkNiftiImageIO.cxx:2167]: Added test.

File: Modules/IO/NIFTI/src/itkNiftiImageIO.cxx (hunk 3 above)
Comment (zivy) [reply Modules/IO/NIFTI/src/itkNiftiImageIO.cxx]: Done.

**Review by dzenanz — APPROVED** [review]

**Review by jhlegarreta — COMMENTED** [review]
Thanks for adding the test @zivy.

I did not realize about the commit message in my previous review, sorry:
- Adding a relevant description as to why the warning is required/convenient/useful would be welcome, as it provides the whole necessary context to understand the change, and eases history navigation (e.g. when blaming commits, navigating history, etc.).
- Minor comment: ending the commit subject with a period is discouraged.

The issue can later be cross-referenced and mark for automatic closing in the PR only.

I know this whole sounds like too strict or pure style-related/unproductive, so I do apologize for it, but helps a lot reviewing/reading commits/PRs.

**Comment by zivy** [comment]
@jhlegarreta, updated commit message to be more informative.

**Review by dzenanz — APPROVED** [review]

**Comment by jhlegarreta** [comment]
While having a look at the coverage today, I realized that it looks like this commit introduced 66 Valgrind warnings :worried: :
https://open.cdash.org/viewDynamicAnalysis.php?buildid=8049748
https://open.cdash.org/viewDynamicAnalysis.php?buildid=8054367

**Comment by blowekamp** [comment]
Addressed in #3525

**Pull request merged**